dask/dataframe/io/parquet/core.py

```diff
@@ -0,0 +1,452 @@
+            filters,
+            categories,
+            index,
+            storage_options,
+            engine,
+            gather_statistics,
+        )
+        return df[columns]
+
+    if columns is not None:
+        columns = list(columns)
+
+    name = "read-parquet-" + tokenize(
+        path,
+        columns,
+        filters,
+        categories,
+        index,
+        storage_options,
+        engine,
+        gather_statistics,
+    )
+
+    if isinstance(engine, str):
+        engine = get_engine(engine)
+
+    fs, fs_token, paths = get_fs_token_paths(
+        path, mode="rb", storage_options=storage_options
+    )
+
+    paths = sorted(paths, key=natural_sort_key)  # numeric rather than glob ordering
+
+    meta, statistics, parts = engine.read_metadata(
+        fs,
+        fs_token,
+        paths,
+        categories=categories,
+        index=index,
+        gather_statistics=gather_statistics,
+    )
+
+    if columns is None:
+        columns = meta.columns
+
+    if not set(columns).issubset(set(meta.columns)):
+        raise KeyError(
+            "The following columns were not found in the dataset %s\n"
+            "The following columns were found %s"
+            % (set(columns) - set(meta.columns), meta.columns)
+        )
+
+    if statistics:
+        parts, statistics = zip(
+            *[
+                (part, stats)
+                for part, stats in zip(parts, statistics)
+                if stats["num-rows"] > 0
+            ]
+        )
+
+        if filters:
+            parts, statistics = apply_filters(parts, statistics, filters)
+
+        out = sorted_columns(statistics)
+
+        if index and out:
+            out = [o for o in out if o["name"] == index]  # only one valid column
+        if index is not False and len(out) == 1:
+            divisions = out[0]["divisions"]
+            index = out[0]["name"]
+        elif index is not False and len(out) > 1:
+            if any(o["name"] == "index" for o in out):
+                [o] = [o for o in out if o["name"] == "index"]
+                divisions = o["divisions"]
+                index = o["name"]
+            else:
+                warnings.warn(
+                    "Multiple sorted columns found: "
+                    + ", ".join(o["name"] for o in out)
+                )
+                divisions = [None] * (len(parts) + 1)
+        else:
+            divisions = [None] * (len(parts) + 1)
+    else:
+        divisions = [None] * (len(parts) + 1)
+
+    if index and index not in columns:
+        columns = list(columns) + [index]
+
+    meta = meta[columns]
+
+    subgraph = {
+        (name, i): (
+            read_parquet_part,
+            engine.read_partition,
+            fs,
+            meta,
+            part["piece"],
+            columns,
+            index,
+            part["kwargs"],
+        )
+        for i, part in enumerate(parts)
+    }
+    if index:
+        meta = meta.set_index(index)
```

Comment:
index of meta should have been set correctly in `engine.read_metadata`, no? This is important, since meta is what the chunk readers will be comparing with.

Reply:
`engine.read_metadata` doesn't determine what the index will be.  It provides us with statistics, we then do our own logic to find out what the index should be (should a sorted column exist), and then we modify things as we like.  The objective is to remove this logic from the respoisibility of the engine.  The engine is just responsible for giving us dtypes, statistics, and pointers to row groups.  We don't want it having to think about anything else.